src/modules/iotjs_module_tls.c

```diff
@@ -14,150 +14,583 @@
+JS_FUNCTION(TlsConstructor) {
+  DJS_CHECK_THIS();
 
   jerry_value_t jtls = JS_GET_THIS();
-  iotjs_tls_t* tls_data = tls_create(jtls);
+  iotjs_tls_t *tls_data = iotjs_tls_create(jtls);
+
+  jerry_value_t opts = jargv[0];
+
+  // Set initial handshake state
+  tls_data->handshake_state = SSL_HANDSHAKE_READY;
 
   int ret = 0;
+
   if ((ret = mbedtls_ctr_drbg_seed(&tls_data->ctr_drbg, mbedtls_entropy_func,
                                    &tls_data->entropy, NULL, 0)) != 0) {
-    return tls_connect_error(&port, &host, &hostname,
-                             "drbg seeding failed, error code: %d", ret);
+    return JS_CREATE_ERROR(COMMON, "drbg seeding failed");
   }
 
-  ret = mbedtls_net_connect(&tls_data->server_fd, iotjs_string_data(&host),
-                            iotjs_string_data(&port), MBEDTLS_NET_PROTO_TCP);
-  if (ret) {
-    return tls_connect_error(&port, &host, &hostname,
-                             "failed to connect to %s:%s, error code: %d",
-                             iotjs_string_data(&host), iotjs_string_data(&port),
-                             ret);
+  // User provided certificate
+  jerry_value_t jca_txt = iotjs_jval_get_property(opts, "ca");
+  if (jerry_value_is_string(jca_txt)) {
+    iotjs_string_t ca_txt = iotjs_jval_as_string(jca_txt);
+    ret = mbedtls_x509_crt_parse(&tls_data->cacert,
+                                 (const unsigned char *)iotjs_string_data(
+                                     &ca_txt),
+                                 (size_t)iotjs_string_size(&ca_txt) + 1);
+
+    iotjs_string_destroy(&ca_txt);
+  } else {
+    // Parse the default certificate
+    ret = mbedtls_x509_crt_parse(&tls_data->cacert,
+                                 (const unsigned char *)SSL_CA_PEM,
+                                 sizeof(SSL_CA_PEM));
   }
 
-  ret = mbedtls_ssl_config_defaults(&tls_data->conf, MBEDTLS_SSL_IS_CLIENT,
-                                    MBEDTLS_SSL_TRANSPORT_STREAM,
-                                    MBEDTLS_SSL_PRESET_DEFAULT);
   if (ret) {
-    return tls_connect_error(&port, &host, &hostname,
-                             "ssl config failed, error code: %d", ret);
+    return JS_CREATE_ERROR(COMMON, "certificate parsing failed");
   }
-
-  mbedtls_ssl_conf_authmode(&tls_data->conf, MBEDTLS_SSL_VERIFY_OPTIONAL);
   mbedtls_ssl_conf_ca_chain(&tls_data->conf, &tls_data->cacert, NULL);
+  jerry_release_value(jca_txt);
+
   mbedtls_ssl_conf_rng(&tls_data->conf, mbedtls_ctr_drbg_random,
                        &tls_data->ctr_drbg);
 
-  ret = mbedtls_ssl_setup(&tls_data->ssl, &tls_data->conf);
+  if ((ret = mbedtls_ssl_config_defaults(&tls_data->conf, MBEDTLS_SSL_IS_CLIENT,
+                                         MBEDTLS_SSL_TRANSPORT_STREAM,
+                                         MBEDTLS_SSL_PRESET_DEFAULT))) {
+    return JS_CREATE_ERROR(COMMON, "SSL Configuration failed.");
+  }
+
+  // if true verifies CAs, must emit error if fails
+  jerry_value_t jreject_unauthorized =
```

Comment:
A test case for this and others seems needed. Can you share your plan to update test_tls.js?